### PR TITLE
test: [POM] Refactor import account e2e tests to use Page Object Model

### LINKED ISSUE #28328: [POM] Migrate import account tests to Page Object Model to reduce flakiness
### What is this about?

Migrate import account tests to Page Object Model to reduce flakiness:
`test/e2e/tests/account/add-account.spec.js`
`test/e2e/tests/account/import-flow.spec.js `

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by chloeYue**
## **Description**

- Migrate e2e tests `test/e2e/tests/account/add-account.spec.js` to TS and Page Object Model, to reduce flakiness.
- Migrate e2e tests `test/e2e/tests/account/import-flow.spec.js` to TS and Page Object Model, to reduce flakiness.
- Deprecate/remove old functions
- Use onboarding related functions designed in page object model, which are more stable, and removed unnecessary delays.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27155?quickstart=1)

## **Related issues**

Fixes: #28328 

## **Manual testing steps**
Check code readability, make sure tests pass.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { strict as assert } from 'assert';
+import { Suite } from 'mocha';
+import { Driver } from '../../webdriver/driver';
+import FixtureBuilder from '../../fixture-builder';
+import { withFixtures, unlockWallet } from '../../helpers';
+import { isManifestV3 } from '../../../../shared/modules/mv3.utils';
+
+describe('Lattice hardware wallet @no-mmi', function (this: Suite) {
```
Comment (chloeYue) [inline test/e2e/tests/hardware-wallets/lattice-connect.spec.ts:8]: This testcase is splited from the test file `import-flow.spec.js` because it is more hardware wallet related. Therefore, we should put it in the same folder with other hardware wallet tests. 
This test need to be converted to page object modal in separate PR.

File: test/e2e/page-objects/flows/send-transaction.flow.ts [hunk 2]
```diff
@@ -40,6 +40,42 @@ export const sendTransaction = async (
...
+ * @param recipientAccount - The recipient account.
+ * @param amount - The amount of the asset to be sent in the transaction.
+ * @param gasfee - The expected transaction gas fee.
+ * @param totalfee - The expected total transaction fee.
+ */
+export const sendTransactionToAccount = async (
+  driver: Driver,
+  recipientAccount: string,
+  amount: string,
+  gasfee: string,
```
Comment (hjetpoluru) [inline test/e2e/page-objects/flows/send-transaction.flow.ts:56]: Minor nit: Spell check is indicating that the correct form is "gasFee"
```suggestion
  gasFee: string,
```
Screenshot 2024-11-07 at 9 10 38 AM

File: test/e2e/page-objects/flows/send-transaction.flow.ts [hunk 3]
```diff
@@ -40,6 +40,42 @@ export const sendTransaction = async (
...
+/**
+ * This function initiates the steps required to send a transaction from the homepage to final confirmation.
+ *
+ * @param driver - The webdriver instance.
+ * @param recipientAccount - The recipient account.
+ * @param amount - The amount of the asset to be sent in the transaction.
+ * @param gasfee - The expected transaction gas fee.
+ * @param totalfee - The expected total transaction fee.
+ */
+export const sendTransactionToAccount = async (
```
Comment (hjetpoluru) [inline test/e2e/page-objects/flows/send-transaction.flow.ts:52]: I see another function `sendTransaction` is there any difference?

File: test/e2e/page-objects/flows/send-transaction.flow.ts (hunk 2 above)
Comment (chloeYue) [reply test/e2e/page-objects/flows/send-transaction.flow.ts]: Thanks for the nice catch ! Addressed in commit https://github.com/MetaMask/metamask-extension/pull/28325/commits/057528a01bdbe692df0eca14ac7a2ff9e9b4d2af

File: test/e2e/page-objects/flows/send-transaction.flow.ts (hunk 3 above)
Comment (chloeYue) [reply test/e2e/page-objects/flows/send-transaction.flow.ts:52]: The parameters are different. `sendTransaction` takes a recipient address as a parameter, while `sendTransactionToAccount` takes an account name as a parameter.
To mitigate this confusion, i have renamed `sendTransaction` to `sendTransactionToAddress`, see commit [057528a](https://github.com/MetaMask/metamask-extension/pull/28325/commits/057528a01bdbe692df0eca14ac7a2ff9e9b4d2af)

**Review by hjetpoluru — COMMENTED** [review]
When reviewing this file, I noticed that this could cause a [potential race condition](https://github.com/MetaMask/metamask-extension/pull/28325/files#diff-2d706063395e4a4db9ab551f1e8a3f1c1defab88b7f615328dfca95867dfda87L77), and this is not part of this PR implementation. I will leave it up to @chloeYue to decide if this change should be implemented here or in another PR.

File: test/e2e/page-objects/flows/send-transaction.flow.ts (hunk 3 above)
Comment (hjetpoluru) [reply test/e2e/page-objects/flows/send-transaction.flow.ts:52]: Ok perfect thanks.

**Comment by chloeYue** [comment]
> When reviewing this file, I noticed that this could cause a [potential race condition](https://github.com/MetaMask/metamask-extension/pull/28325/files#diff-2d706063395e4a4db9ab551f1e8a3f1c1defab88b7f615328dfca95867dfda87L77), and this is not part of this PR implementation. I will leave it up to @chloeYue to decide if this change should be implemented here or in another PR.

Hi @hjetpoluru, could you check the link ? when i open it, it shows the whole file changes?

File: test/e2e/tests/confirmations/transactions/erc20-token-send-redesign.spec.ts [hunk 4]
```diff
@@ -144,8 +144,8 @@ async function createWalletInitiatedTransactionAndAssertDetails(
   await sendToPage.fillRecipient('0x2f318C334780961FB129D2a6c30D0763d9a5C970');
   await sendToPage.fillAmount('1');
 
-  await sendToPage.click_assetPickerButton();
-  await sendToPage.click_secondTokenListButton();
+  await sendToPage.clickAssetPickerButton();
```
Comment (hjetpoluru) [inline test/e2e/tests/confirmations/transactions/erc20-token-send-redesign.spec.ts:147]: Nice catch!

File: test/e2e/page-objects/pages/send/send-token-page.ts [hunk 5]
```diff
@@ -60,11 +50,13 @@ class SendTokenPage {
...
+  async clickAssetPickerButton() {
+    await this.driver.clickElement(this.assetPickerButton);
+  }
+
+  async clickSecondTokenListButton() {
+    const elements = await this.driver.findElements(this.tokenListButton);
+    await elements[1].click();
   }
 
   async fillAmount(amount: string): Promise<void> {
```
Comment (hjetpoluru) [inline test/e2e/page-objects/pages/send/send-token-page.ts:62]: @chloeYue, the link takes to the file. Hope this message helps, this could cause race conditional and this is not part of this PR implementation.

Comment (chloeYue) [reply test/e2e/page-objects/pages/send/send-token-page.ts:62]: Thanks , this function indeed seems anti pattern, i will take a note and address it in separate PR.

Comment (hjetpoluru) [reply test/e2e/page-objects/pages/send/send-token-page.ts:62]: Thanks @chloeYue

**Review by hjetpoluru — DISMISSED** [review]
LGTM nice effort.

File: test/e2e/page-objects/flows/onboarding.flow.ts [hunk 6]
```diff
@@ -43,17 +43,20 @@ export const createNewWalletOnboardingFlow = async (
...
- * @param [options.password] - The password to use.
+ * @param [options.seedPhrase] - The seed phrase to import. Defaults to E2E_SRP.
+ * @param [options.password] - The password to use. Defaults to WALLET_PASSWORD.
+ * @param [options.fillSrpWordByWord] - Whether to fill the SRP word by word. Defaults to false.
  */
 export const importSRPOnboardingFlow = async ({
   driver,
   seedPhrase = E2E_SRP,
   password = WALLET_PASSWORD,
+  fillSrpWordByWord = false,
```
Comment (seaona) [inline test/e2e/page-objects/flows/onboarding.flow.ts:54]: nice!

File: test/e2e/helpers.js [hunk 7]
```diff
@@ -1234,6 +1133,10 @@ async function initBundler(bundlerServer, ganacheServer, usePaymaster) {
   }
 }
 
+/**
+ * @deprecated Please use page object functions in `pages/account-list-page`.
+ * @param driver
+ */
```
Comment (seaona) [inline test/e2e/helpers.js:1139]: nice!

File: test/e2e/page-objects/flows/send-transaction.flow.ts [hunk 8]
```diff
@@ -10,15 +10,15 @@ import SnapSimpleKeyringPage from '../pages/snap-simple-keyring-page';
...
  */
-export const sendTransaction = async (
+export const sendTransactionToAddress = async (
   driver: Driver,
   recipientAddress: string,
   amount: string,
-  gasfee: string,
-  totalfee: string,
+  gasFee: string,
+  totalFee: string,
```
Comment (seaona) [inline test/e2e/page-objects/flows/send-transaction.flow.ts:21]: I think maybe here we could do the same operation of using an object instead of single params :thinking: 
Given that this it is only used in 2 specs, the change won't be too big yet. If we want to do this later in time, where more tests use this, it would be more challenging. So maybe it's worth considering. 
What do you think?

File: test/e2e/page-objects/flows/send-transaction.flow.ts [hunk 9]
```diff
@@ -36,7 +36,43 @@ export const sendTransaction = async (
...
+ * @param amount - The amount of the asset to be sent in the transaction.
+ * @param gasFee - The expected transaction gas fee.
+ * @param totalFee - The expected total transaction fee.
+ */
+export const sendTransactionToAccount = async (
+  driver: Driver,
+  recipientAccount: string,
+  amount: string,
+  gasFee: string,
+  totalFee: string,
```
Comment (seaona) [inline test/e2e/page-objects/flows/send-transaction.flow.ts:57]: the same comment as above, about using an object as params. In this case, this is also only used in 2 specs now, so if we want to do this change, I think it is a good moment now

File: test/e2e/page-objects/pages/account-list-page.ts [hunk 10]
```diff
@@ -89,6 +123,9 @@ class AccountListPage {
     await this.driver.clickElement(this.createAccountButton);
     await this.driver.clickElement(this.addEthereumAccountButton);
     await this.driver.fill(this.accountNameInput, customLabel);
+    // needed to mitigate a race condition with the state update
+    // there is no condition we can wait for in the UI
```
Comment (seaona) [inline test/e2e/page-objects/pages/account-list-page.ts:127]: :+1: thanks for adding the comment! Make sense to me

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+          tag: 'button',
+        });
+        await driver.findClickableElement(
+          '[data-testid="hardware-connect-close-btn"]',
+        );
+        await driver.clickElement('[data-testid="connect-lattice-btn"]');
+        await driver.findClickableElement({
+          text: 'Continue',
+          tag: 'button',
+        });
```
Comment (seaona) [inline test/e2e/tests/hardware-wallets/lattice-connect.spec.ts:40]: I think we can remove this, as it's implicit in the next step

```suggestion
```

**Review by seaona — DISMISSED** [review]
awesome work @chloeYue :fire: overall looks good! I've just added a couple of small comments, but feel free to take them or leave them

File: test/e2e/tests/hardware-wallets/lattice-connect.spec.ts (hunk 11 above)
Comment (chloeYue) [reply test/e2e/tests/hardware-wallets/lattice-connect.spec.ts]: Make sense, done

File: test/e2e/page-objects/flows/send-transaction.flow.ts (hunk 9 above)
Comment (chloeYue) [reply test/e2e/page-objects/flows/send-transaction.flow.ts]: Thanks for the suggestion, addressed in [0aa97f0](https://github.com/MetaMask/metamask-extension/pull/28325/commits/0aa97f0cfa2bba99b73f4933263a3ba03c915340)

**Review by seaona — APPROVED** [review]

**Review by hjetpoluru — APPROVED** [review]
